lib/index.js

```diff
@@ -302,19 +302,17 @@ internals.isObjectId = joiObject => {
   return false;
 };
 
-internals.mongooseValidateWrapper = async (originalJoiSchema, value) => {
+internals.mongooseValidateWrapper = (originalJoiSchema, value) => {
   var joiSchema = Hoek.clone(originalJoiSchema);
 
   if (joiSchema._isObjectId) {
-    joiSchema = Joi.alternatives([
+    joiSchema = Joi.alternatives(
       joiSchema,
-      Joi.object().type(internals.mongoose.Types.ObjectId)
-    ]);
+      Joi.object().instance(internals.mongoose.Types.ObjectId)
+    );
   }
 
-  try {
-    await Joi.validate(value, joiSchema, internals.joiGlobalOptions);
-  } catch (err) {
+  if (joiSchema.validate(value, internals.joiGlobalOptions).error) {
```

Comment:
`joiSchema.validate()` returns an object, while `joiSchema.validateAsync()` will return a promise. I'm not sure about which was the expected behaviour and the reason behind the asynchronicity of this method. When returning an object is fine, then it should be possible to remove the `await` keywords from the tests.

Reply:
There are some joi validators which can be async. For example, check out: https://hapi.dev/family/joi/?v=16.1.8#anyexternalmethod-description

If we can determine whether or not an async validator is used, we choose which method to call. This might help: https://github.com/hapijs/joi/blob/c97f2d55218deb2e0b86a6219936486694d76ffc/lib/validator.js#L27

Reply:
It's just this which I think needs reverting for the reasons specified above.

Reply:
I'm taking a look at it. I will let you know soon if I can solve this as well.

Reply:
What about the current solution ? It's a pretty naive approach, but it seems to work. Am I missing any important test case about it ?

Reply:
That looks like it works! Good job! There's one test case missing: when the async validation fails.

Reply:
Ok, this is covered as well now.